### PR TITLE
ADR records for multiple queuing systems and bunny-mock

### PR DESCRIPTION
**Pull request by dwhenry**
https://trello.com/c/bFnfSdqI/167-implement-message-requeuing-when-using-the-new-index

### CODE REVIEW

File: doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# Decision record: use of both RabbitMQ (bunny) and Redis (Sidekiq) queuing systems
+
+## Context
+
+In order to simplify the search infrasture we are moving to using the publishing API
```
Comment (brenetic) [inline doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md]: `infrastructure`

File: doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+as the source of data. This allows us to reduce the amount of code duplication across
+projects, as current each publishing app has to implement a write interface into rummager.
```
Comment (brenetic) [inline doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md]: `currently`

File: doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+We expect this will give us a number of improvements, including:
+
+* rebuild the search index more easily
+* reduce the load on the publishing API - rummager current needs to retrieve tagging data
```
Comment (brenetic) [inline doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md]: `currently`

File: doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  from the publishing API when it indexes documents
+
+The publishing API currently uses rabbit MQ to tell downstream apps that a new publishing
+event has occurred, with the key specifying teh publishing app and the publishing event
```
Comment (brenetic) [inline doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md]: `the`

File: doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+type.
+
+Rummager currently processes `link` events from the rabbit MQ stream, this works by reading
+the message off the rabbitMQ stream, doing a small about of manipulation and then creating
```
Comment (brenetic) [inline doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md]: `amount`

File: doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+* Concurrency
+* Failure/Queue side tracking
+* Retry logic
+
+
+## Decision
+
+We initially looked at two options, either reading the data from the rabbit MQ stream and
+processing inplace, or writting the message to a Sidekiq worker and then have it perform
```
Comment (brenetic) [inline doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md]: `writing`

File: doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+the require processing.
```
Comment (brenetic) [inline doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md]: `required`

File: doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+While it is possible to have implement all the above feature using just the rabbit MQ system
```
Comment (brenetic) [inline doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md]: ~~have~~

`features`

File: doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+## Status
+
+Accepted.
+
+## Consequences
+
+This means we have additional points of failure in the system, as we are reading from one queue
+and writing to second one. This is especially important when debugging and we have a clear
+diagram for the data flow through the system as well as a documented way to determin the status
```
Comment (brenetic) [inline doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md]: `determine`

File: doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+of a message within the system.
+
+In addition to this it is important to differentiate the different components and ensure that
+new developers are clear when gems/libraries are assocaited with which queuing system.
```
Comment (brenetic) [inline doc/arch/adr-001-use-of-both-rabbitmq-and-sidekiq-queues.md]: ~~when~~  => `which`

`associated`

**Review by brenetic — CHANGES_REQUESTED** [review]

**Review by MatMoore — COMMENTED** [review]
Got half way through reviewing before being distracted, so some of mine probably duplicate Brendan's. Other than a few typos, it looks good to me.

**Comment by dwhenry** [comment]
@MatMoore @brenetic I have now completed the above updates.

I have removed the second ADR (it was just a placeholder) and will look to add that in a separate PR

**Review by brenetic — APPROVED** [review]

**Pull request merged**